vmm/src/serial_manager.rs

```diff
@@ -0,0 +1,245 @@
+// Copyright © 2021 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use crate::config::ConsoleOutputMode;
+use crate::device_manager::PtyPair;
+#[cfg(target_arch = "aarch64")]
+use devices::legacy::Pl011;
+#[cfg(target_arch = "x86_64")]
+use devices::legacy::Serial;
+use libc::EFD_NONBLOCK;
+use std::fs::File;
+use std::io::Read;
+use std::os::unix::io::{AsRawFd, FromRawFd};
+use std::panic::AssertUnwindSafe;
+use std::sync::{Arc, Mutex};
+use std::{io, result, thread};
+use thiserror::Error;
+use vmm_sys_util::eventfd::EventFd;
+
+#[derive(Debug, Error)]
+pub enum Error {
+    /// Cannot clone File.
+    #[error("Error cloning File: {0}")]
+    FileClone(#[source] io::Error),
+
+    /// Cannot create epoll context.
+    #[error("Error creating epoll context: {0}")]
+    Epoll(#[source] io::Error),
+
+    /// Cannot handle the VM input stream.
+    #[error("Error handling VM input: {0:?}")]
+    ReadInput(#[source] io::Error),
+
+    /// Cannot queue input to the serial device.
+    #[error("Error queuing input to the serial device: {0}")]
+    QueueInput(#[source] vmm_sys_util::errno::Error),
+
+    /// Cannot make the file descriptor non-blocking.
+    #[error("Error making input file descriptor non-blocking: {0}")]
+    SetNonBlocking(#[source] io::Error),
+
+    /// Cannot create EventFd.
+    #[error("Error creating EventFd: {0}")]
+    EventFd(#[source] io::Error),
+
+    /// Cannot spawn SerialManager thread.
+    #[error("Error spawning SerialManager thread: {0}")]
+    SpawnSerialManager(#[source] io::Error),
+}
+pub type Result<T> = result::Result<T, Error>;
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+#[repr(u64)]
+pub enum EpollDispatch {
+    File = 0,
+    Kill = 1,
+    Unknown,
+}
+
+impl From<u64> for EpollDispatch {
+    fn from(v: u64) -> Self {
+        use EpollDispatch::*;
+        match v {
+            0 => File,
+            1 => Kill,
+            _ => Unknown,
+        }
+    }
+}
+
+pub struct SerialManager {
+    #[cfg(target_arch = "x86_64")]
+    serial: Arc<Mutex<Serial>>,
+    #[cfg(target_arch = "aarch64")]
+    serial: Arc<Mutex<Pl011>>,
+    epoll_file: File,
+    in_file: File,
+    kill_evt: EventFd,
+    handle: Option<thread::JoinHandle<()>>,
+}
+
+impl SerialManager {
+    pub fn new(
+        #[cfg(target_arch = "x86_64")] serial: Arc<Mutex<Serial>>,
+        #[cfg(target_arch = "aarch64")] serial: Arc<Mutex<Pl011>>,
+        pty_pair: Option<Arc<Mutex<PtyPair>>>,
+        mode: ConsoleOutputMode,
+    ) -> Result<Option<Self>> {
+        let in_file = match mode {
+            ConsoleOutputMode::Pty => {
+                if let Some(pty_pair) = pty_pair {
+                    pty_pair
+                        .lock()
+                        .unwrap()
+                        .main
+                        .try_clone()
+                        .map_err(Error::FileClone)?
+                } else {
+                    return Ok(None);
```

Comment:
Shouldn't this be treated as an error? I mean if the mode is Pty and we haven't been given an actual PtyPair, this should raise a flag.

Reply:
I think after a restore the `PtyPair` is None.

Reply:
@rbradford I thought `PtyPair` should still be set assuming the `serial_pty` was correctly passed though from the vmm?

@sboeuf It could be an error. I ended up being quite forgiving on creating a `SerialManager` given it is only handling input and it may be reasonable to have a pty device be removable in the future.

Reply:
Fair enough :)